File: example/joint_teleop_over_wan.cpp
/**
 * @example joint_teleop_over_wan.cpp
 * @brief joint-space robot-robot teleoperation over WAN (Wide Area Network, i.e. Internet)
 * connection.
 * @copyright Copyright (C) 2016-2025 Flexiv Ltd. All Rights Reserved.
 * @author Flexiv
 */

#include <flexiv/tdk/joint_teleop_wan.hpp>

#include <spdlog/spdlog.h>

#include <getopt.h>
#include <iostream>

namespace {
const std::vector<double> kJointStiffnessRatio = {0.02, 0.02, 0.02, 0.02, 0.02, 0.02, 0.02};
constexpr double kLastJointShapedInertia = 0.05;
}

void PrintHelp()
{
    // clang-format off
    std::cout << "Usage: sudo ./test_joint_teleop_over_wan [-s serial_num] [-r server/client] [-i ip] [-p port]" << std::endl;
    std::cout << "  -s  --serial-number    Serial number of the local robot." << std::endl;
    std::cout << "  -r  --tcp-role         Role in the TCP connection, [server] or [client]." << std::endl;
    std::cout << "  -i  --ip               Public IPv4 address of the TCP server machine." << std::endl;
    std::cout << "  -p  --port             Listening port of the TCP server machine." << std::endl;
    // clang-format on
}

const struct option kLongOptions[] = {
    // clang-format off
    {"serial number",               required_argument,  0, 's'},
    {"tcp role",                    required_argument,  0, 'r'},
    {"public ipv4 address",         required_argument,  0, 'i'},
    {"port",                        required_argument,  0, 'p'},
    {"lan whitelist ip",            optional_argument,  0, 'l'},
    {"wan whitelist ip",            optional_argument,  0, 'w'},
    {0,                             0,                  0,  0 }
    // clang-format on
};

int main(int argc, char* argv[])
{
    std::string local_sn, tcp_role, public_server_ip, lan_ip, wan_ip;
    unsigned int server_port = 0;
    std::vector<std::string> lan_interface_whitelist {};
    std::vector<std::string> wan_interface_whitelist {};

    int opt = 0;
    while ((opt = getopt_long_only(argc, argv, "s:r:i:p:l:w:", kLongOptions, nullptr)) != -1) {
        switch (opt) {
            case 's':
                local_sn = std::string(optarg);
                break;
            case 'r':
                tcp_role = std::string(optarg);
                break;
            case 'i':
                public_server_ip = std::string(optarg);
                break;
            case 'p':
                try {
                    server_port = std::stoi(optarg);
                } catch (...) {
                    spdlog::error("Invalid port number: {}", optarg);
                    return 1;
                }
                break;
            case 'l':
                lan_ip = std::string(optarg);
                lan_interface_whitelist.push_back(lan_ip);
                break;
            case 'w':
                wan_ip = std::string(optarg);
                wan_interface_whitelist.push_back(wan_ip);
                break;
            default:
                PrintHelp();
                return 1;
        }
    }
    if (local_sn.empty() || tcp_role.empty() || public_server_ip.empty() || server_port == 0) {
        PrintHelp();
        return 1;
    }
    if (lan_interface_whitelist.empty() || wan_interface_whitelist.empty()) {
        spdlog::warn("LAN or WAN whitelist is not provided, will search all network interfaces.");
    }

    // Whether this is a TCP server or client
    bool is_tcp_server;
    if (tcp_role == "server") {
        is_tcp_server = true;
    } else if (tcp_role == "client") {
        is_tcp_server = false;
    } else {
        spdlog::error("Valid inputs for [-r] are: server, client");
        return 1;
    }

    // Network configuration
    flexiv::tdk::NetworkCfg network_cfg;
    network_cfg.is_tcp_server = is_tcp_server;
    network_cfg.public_ipv4_address = public_server_ip;
    network_cfg.listening_port = server_port;
    network_cfg.lan_interface_whitelist = lan_interface_whitelist;
    network_cfg.wan_interface_whitelist = wan_interface_whitelist;

    spdlog::info("Robot serial number: {}", local_sn);
    spdlog::info("Start as {} on {}:{}", tcp_role, public_server_ip, server_port);

    try {
        // Instantiate robot node
        flexiv::tdk::JointTeleopWAN joint_teleop(local_sn, network_cfg);

        // Run initialization sequence
        joint_teleop.Init();

        if (is_tcp_server) {
            // Set 20 degrees soft limit for only one side of teleoperation
            joint_teleop.SetSoftLimit(20.0);

            // Server stays at current pose
            joint_teleop.SyncPose(false, {});
        } else {
            // Client syncs pose with server
            joint_teleop.SyncPose(true);
        }

        // Enable inertia shaping for the last joint
        std::vector<std::pair<bool, double>> shaped_joint_inertia;
        shaped_joint_inertia.resize(joint_teleop.DoF(), {false, 1.0});
        shaped_joint_inertia.back() = {true, kLastJointShapedInertia};
        joint_teleop.SetInertiaShaping(shaped_joint_inertia);

        // Start control loop
        joint_teleop.Start();

        // Set joint impedance
        joint_teleop.SetJointImpedance(kJointStiffnessRatio);

        // Block until faulted
        bool last_pedal_input = false;
        while (!joint_teleop.fault()) {
            // Server is activated by pedal, client will auto sync the activation
            if (is_tcp_server) {
                bool pedal_input = joint_teleop.digital_inputs()[0];
                if (pedal_input != last_pedal_input) {
                    joint_teleop.Activate(pedal_input);
                    last_pedal_input = pedal_input;
                }
            }
            std::this_thread::sleep_for(std::chrono::milliseconds(100));
        }

        // Fault occurred, stop the teleoperation
        joint_teleop.Stop();

    } catch (const std::exception& e) {
        spdlog::error(e.what());
        return 1;
    }

    return 0;
}

File: example_py/joint_teleop_over_wan.py
#!/usr/bin/env python3
"""
joint_teleop_over_wan.py

Example usage of joint-space robot-robot teleoperation over WAN (Wide Area Network) connection.
Monitors pedal input from robot connected to the server machine to activate/deactivate teleoperation.
"""

import sys
import argparse
import time
import spdlog
# pip install flexivtdk
import flexivtdk

# Constants
JOINT_STIFFNESS_RATIO = [0.02, 0.02, 0.02, 0.02, 0.02, 0.02, 0.02]
LAST_JOINT_SHAPED_INERTIA = 0.05

# Logger setup
logger = spdlog.ConsoleLogger("Example")

def print_help():
    """Print usage information"""
    help_text = """Usage: python joint_teleop_over_wan.py [-s serial_num] [-r server/client] [-i ip] [-p port]
  -s, --serial-number    <required> Serial number of the local robot.
  -r, --tcp-role         <required> Role in the TCP connection, [server] or [client].
  -i, --ip               <required> Public IPv4 address of the TCP server machine.
  -p, --port             <required> Listening port of the TCP server machine.
  -l, --lan-whitelist-ip <optional> The ip address of the network card connected to the robot control box.
  -w, --wan-whitelist-ip <optional> The ip address of the network card connected to the Internet."""
    print(help_text)

def main():
    parser = argparse.ArgumentParser(
        description="Joint Teleoperation over WAN",
        add_help=False
    )
    
    parser.add_argument('-s', '--serial-number', required=True, 
                       help='Serial number of the local robot')
    parser.add_argument('-r', '--tcp-role', required=True,
                       choices=['server', 'client'],
                       help='Role in the TCP connection: server or client')
    parser.add_argument('-i', '--ip', required=True,
                       help='Public IPv4 address of the TCP server machine')
    parser.add_argument('-p', '--port', type=int, required=True,
                       help='Listening port of the TCP server machine')
    parser.add_argument('-l', '--lan-whitelist-ip',
                       help='IP address of the network card connected to the robot')
    parser.add_argument('-w', '--wan-whitelist-ip',
                       help='IP address of the network card connected to the Internet')
    
    # If no arguments provided, show help
    if len(sys.argv) == 1:
        print_help()
        sys.exit(1)
    
    try:
        args = parser.parse_args()
    except SystemExit:
        print_help()
        sys.exit(1)
    
    # Validate required arguments
    if not args.serial_number or not args.tcp_role or not args.ip or not args.port:
        print_help()
        sys.exit(1)
    
    # Setup network configuration
    lan_interface_whitelist = []
    wan_interface_whitelist = []
    
    if args.lan_whitelist_ip:
        lan_interface_whitelist.append(args.lan_whitelist_ip)
    if args.wan_whitelist_ip:
        wan_interface_whitelist.append(args.wan_whitelist_ip)
    
    if not lan_interface_whitelist or not wan_interface_whitelist:
        logger.warn("LAN or WAN whitelist is not provided, will search all network interfaces.")
    
    # Determine TCP role
    is_tcp_server = args.tcp_role == 'server'
    if args.tcp_role not in ['server', 'client']:
        logger.error("Valid inputs for --tcp-role are: server, client")
        sys.exit(1)
    
    # Network configuration
    network_cfg = flexivtdk.NetworkCfg()
    network_cfg.is_tcp_server = is_tcp_server
    network_cfg.public_ipv4_address = args.ip
    network_cfg.listening_port = args.port
    network_cfg.lan_interface_whitelist = lan_interface_whitelist
    network_cfg.wan_interface_whitelist = wan_interface_whitelist
    
    logger.info(f"Robot serial number: {args.serial_number}")
    logger.info(f"Start as {args.tcp_role} on {args.ip}:{args.port}")
    
    try:
        # Create teleop control interface instance
        joint_teleop = flexivtdk.JointTeleopWAN(args.serial_number, network_cfg)
        
        # Run initialization sequence
        joint_teleop.Init()
        
        if is_tcp_server:
            # Set 20 degrees soft limit for only one side of teleoperation
            joint_teleop.SetSoftLimit(20.0)
            
            # Server stays at current pose
            joint_teleop.SyncPose(False, [])
        else:
            # Client syncs pose with server
            joint_teleop.SyncPose(True)
        
        # Enable inertia shaping for the last joint
        shaped_joint_inertia = []
        dof = joint_teleop.DoF()
        for i in range(dof):
            if i == dof - 1:  # Last joint
                shaped_joint_inertia.append((True, LAST_JOINT_SHAPED_INERTIA))
            else:
                shaped_joint_inertia.append((False, 1.0))
        
        joint_teleop.SetInertiaShaping(shaped_joint_inertia)
        
        # Start control loop
        joint_teleop.Start()
        
        # Set joint impedance
        joint_teleop.SetJointImpedance(JOINT_STIFFNESS_RATIO)
        
        # Block until faulted
        last_pedal_input = False
        while not joint_teleop.fault():
            # Server is activated by pedal, client will auto sync the activation
            if is_tcp_server:
                pedal_input = joint_teleop.digital_inputs()[0]
                if pedal_input != last_pedal_input:
                    joint_teleop.Activate(pedal_input)
                    last_pedal_input = pedal_input
            # 100ms sleep
            time.sleep(0.1)  
        
        # Fault occurred, stop the teleoperation
        joint_teleop.Stop()
        
    except Exception as e:
        logger.error(str(e))
        sys.exit(1)

if __name__ == "__main__":
    main()



